### PR TITLE
Developing the nearest neighbor matching and stitching

### PR DESCRIPTION
**Pull request by kdorheim**
@abigailsnyder please see  [here](https://rpubs.com/kdorheim/740838) for some notes about what I've done here. Recall that this PR is not going to be merged into the main branch but contains R development of the nearest neighbor matching/stitching to expedite getting data for the April 12th meeting. 

@abigailsnyder and @claudiatebaldi 
The rpub has more details about what is going on and what the next steps might be but here are the highlights! 

We have the ability to match target data with archive data as illustrated here. 
image

which can be used to stitch together global mean temp! the grey line in the raw CanESM data and the red line is our stiched together synthetic ts using `ssp126` and `ssp585` data. 

image

Problems 
* right now there is no great way to pick between equivalent matches 
* the stitching falls apart during the historical to future scenario transition :(

### CODE REVIEW

**Comment by claudiatebaldi** [comment]
What is it about the historical-scenario hand-over that makes it difficult to match?

File: notebooks/stitches_dev/nearest_neighboor_matching.R [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+#   fx: a single value of the target fx value
+#   dx: a single value of the target dx value 
+#   data: a data frame of the archive fx and dx values 
+# Return: a dt
+internal_dist <- function(fx, dx, data){
+  
+  #message(data)
+  # calculate the elucian distance between the target point 
+  # and the archive obversations 
+  dist <- unlist(sqrt((data['dx'] - dx)^2 + (data['fx'] - fx)^2))
```
Comment (kdorheim) [inline notebooks/stitches_dev/nearest_neighboor_matching.R:16]: @abigailsnyder this is where we will want to save a copy of the value distance vs the rate of change distance. then add it as columns to the out data frame below.

File: notebooks/stitches_dev/nearest_neighboor_matching.R [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+match_nearest_neighboor <- function(target_data, archive_data){
+  
+# Check the inputs of the functions 
+req_cols <- c("experiment", "variable", "ensemble", "start_yr", "end_yr", "fx", "dx")  
+assert_that(has_name(which = req_cols, x = archive_data))
+req_cols <- c("start_yr", "end_yr", "fx", "dx")
+assert_that(has_name(which = req_cols, x = target_data))
+
+
+# For every entry in the target data frame find its nearest neighboor from the archive data.
```
Comment (kdorheim) [inline notebooks/stitches_dev/nearest_neighboor_matching.R:51]: might want to replace the `internal_dist` call here with a function from an R package, knn n = 1, the thing I was struggling with when I was looking into the prebuilt functions was a way to match from different data sets or between different variable/columns of data.

File: notebooks/stitches_dev/sitching_functions.R [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  data_list <- lapply(1:nrow(match), 
+                      FUN = function(i){
+                        # select a single row of matched data, this is akward but not using a lapply here 
+                        # because it will transform the int data into strings 
+                        m <- match[i, ]
+                        yr <- m$`target-start_yr`:m$`target-end_yr`
+                        
+                        select_yrs <- m$`archive-start_yr`:m$`archive-end_yr`
+                        
+                        if(any(select_yrs < 2015)){
```
Comment (kdorheim) [inline notebooks/stitches_dev/sitching_functions.R:41]: this if the statement was how I was trying to deal with the historical data in the archive, however, if the archive data is stitched historical-future then the if statement can be dropped!

File: notebooks/stitches_dev/nearest_neighboor_matching.R (hunk 2 above)
Comment (kdorheim) [reply notebooks/stitches_dev/nearest_neighboor_matching.R:51]: CT suggested the package class might be useful for this

**Comment by kdorheim** [comment]
Priority Issues 
* is the nearest neighbor matching really working? It looks like it might be missing the actual nearest neighbor
     * incorporate the nearest time element into the distance space?  
* address the gap in the stitched emualted data by concatenating historical-future data in the tgav data archive

**Review by abigailsnyder — APPROVED** [review]

**Pull request merged**